File: solutions/java/0567.java
class Solution {
  public boolean checkInclusion(String s1, String s2) {
    int[] count1 = new int[26];
    int required = s1.length();

    for (char c : s1.toCharArray())
      ++count1[c - 'a'];

    for (int r = 0; r < s2.length(); ++r) {
      if (--count1[s2.charAt(r) - 'a'] >= 0)
        --required;
      if (r >= s1.length())
        if (++count1[s2.charAt(r - s1.length()) - 'a'] > 0)
          ++required;
      if (required == 0)
        return true;
    }

    return false;
  }
}

File: solutions/python3/0567.py
class Solution:
    def checkInclusion(self, s1: str, s2: str) -> bool:
        count1 = collections.Counter(s1)
        required = len(s1)

        for r, c in enumerate(s2):
            count1[c] -= 1
            if count1[c] >= 0:
                required -= 1
            if r >= len(s1):
                count1[s2[r - len(s1)]] += 1
                if count1[s2[r - len(s1)]] > 0:
                    required += 1
            if required == 0:
                return True

        return False
